### PR TITLE
KOGITO-7077: Add support for collections in dataindex

### PR DESCRIPTION
**Pull request by nmirasch**
jira: https://issues.redhat.com/browse/KOGITO-7077

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

- for <b>pull request checks</b>  
  Please add comment: <b>Jenkins retest this</b>

- for a <b>specific pull request check</b>  
  Please add comment: <b>Jenkins (re)run [kogito-apps|kogito-examples] tests</b>

- for <b>quarkus branch checks</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins run quarkus-branch</b>

- for a <b>quarkus branch specific check</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins (re)run [kogito-apps|kogito-examples] quarkus-branch</b>

- for <b>quarkus main checks</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins run quarkus-main</b>

- for a <b>specific quarkus main check</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins (re)run [kogito-apps|kogito-examples] quarkus-branch</b>

- for <b>native checks</b>  
  Run native checks  
  Please add comment: <b>Jenkins run native</b>

- for a <b>specific native check</b>  
  Run native checks 
  Please add comment: <b>Jenkins (re)run [kogito-apps|kogito-examples] native</b>

- for <b>mandrel checks</b>  
  Run native checks against Mandrel image
  Please add comment: <b>Jenkins run mandrel</b>

- for a <b>specific mandrel check</b>  
  Run native checks against Mandrel image  
  Please add comment: <b>Jenkins (re)run [kogito-apps|kogito-examples] mandrel</b>

</details>

### CODE REVIEW

File: data-index/data-index-service/data-index-service-common/src/main/java/org/kie/kogito/index/graphql/JsonPropertyDataFetcher.java [hunk 1]
```diff
@@ -41,7 +45,22 @@ public Object get(DataFetchingEnvironment environment) {
                 switch (value.getNodeType()) {
                     case OBJECT:
                     case POJO:
+                        return value;
                     case ARRAY:
+                        if (!value.isNull() && !value.isEmpty()) {
+                            ObjectMapper mapper = new ObjectMapper();
```
Comment (cristianonicolai) [inline data-index/data-index-service/data-index-service-common/src/main/java/org/kie/kogito/index/graphql/JsonPropertyDataFetcher.java]: I guess the mapper and TypeReference objects could be at the instance level for reuse.

File: data-index/data-index-service/data-index-service-mongodb/src/test/resources/travels-mongo.proto [hunk 2]
```diff
@@ -38,6 +38,8 @@ message Traveller {
     optional string firstName = 3;
     optional string lastName = 4;
     optional string nationality = 5;
+    repeated Address locations = 6;
+    repeated string aliases=7;
```
Comment (cristianonicolai) [inline data-index/data-index-service/data-index-service-mongodb/src/test/resources/travels-mongo.proto]:
```suggestion
    repeated string aliases = 7;
```

File: data-index/data-index-service/data-index-service-common/src/test/resources/travels.proto [hunk 3]
```diff
@@ -57,7 +57,11 @@ message Traveller {
 	/* @Field(index = Index.NO, store = Store.YES) @SortableField */ 
 	optional string lastName = 4; 
 	/* @Field(index = Index.NO, store = Store.YES) @SortableField */ 
-	optional string nationality = 5; 
+	optional string nationality = 5;
+		/* @Field(index = Index.NO, store = Store.YES) */
```
Comment (cristianonicolai) [inline data-index/data-index-service/data-index-service-common/src/test/resources/travels.proto]:
```suggestion
	/* @Field(index = Index.NO, store = Store.YES) */
```

File: data-index/data-index-service/data-index-service-common/src/test/resources/travels.proto [hunk 4]
```diff
@@ -57,7 +57,11 @@ message Traveller {
...
+	repeated Address locations = 6;
+		/* @Field(index = Index.NO, store = Store.YES) */
```
Comment (cristianonicolai) [inline data-index/data-index-service/data-index-service-common/src/test/resources/travels.proto]:
```suggestion
	/* @Field(index = Index.NO, store = Store.YES) */
```

**Review by cristianonicolai — COMMENTED** [review]
@nmirasch great, just some minor comments

**Comment by radtriste** [comment]
jenkins retest this

**Comment by cristianonicolai** [comment]
jenkins please retest this

**Review by cristianonicolai — APPROVED** [review]

File: data-index/data-index-service/data-index-service-common/src/main/java/org/kie/kogito/index/graphql/JsonPropertyDataFetcher.java [hunk 5]
```diff
@@ -15,17 +15,28 @@
...
 public class JsonPropertyDataFetcher implements DataFetcher {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(JsonPropertyDataFetcher.class);
+    private ObjectMapper mapper = new ObjectMapper();
+    private TypeReference stringListTypeRef = new TypeReference<List<String>>() {
+    };
+    private TypeReference numberListTypeRef = new TypeReference<List<Number>>() {
+    };
+    private TypeReference booleanListTypeRef = new TypeReference<List<Boolean>>() {
+    };
```
Comment (radtriste) [inline data-index/data-index-service/data-index-service-common/src/main/java/org/kie/kogito/index/graphql/JsonPropertyDataFetcher.java]: can we have those static ?

File: data-index/data-index-service/data-index-service-common/src/main/java/org/kie/kogito/index/graphql/query/AbstractInputObjectTypeMapper.java [hunk 6]
```diff
@@ -71,4 +74,16 @@ protected Map<String, GraphQLType> getAdditionalTypes() {
...
         return (GraphQLInputObjectType) schema.getType(name);
     }
+
+    protected String resolveBaseTypeName(GraphQLOutputType graphQLOutputType) {
+        if (graphQLOutputType instanceof GraphQLList) {
+            String baseType = ((GraphQLNamedType) graphQLOutputType.getChildren().get(0)).getName();
+            if ("String".equals(baseType)) {
+                return "StringArray";
+            }
+            return ((GraphQLNamedType) graphQLOutputType.getChildren().get(0)).getName();
```
Comment (radtriste) [inline data-index/data-index-service/data-index-service-common/src/main/java/org/kie/kogito/index/graphql/query/AbstractInputObjectTypeMapper.java]:
```suggestion
            return baseType;
```
?

File: data-index/data-index-service/data-index-service-common/src/main/java/org/kie/kogito/index/graphql/query/GraphQLOrderByTypeMapper.java [hunk 7]
```diff
@@ -49,39 +50,41 @@ public String getTypeName(GraphQLObjectType type) {
...
+                switch (field.getName()) {
+                    //Skip id, multi instances not sortable
+                    case PROCESS_INSTANCES_DOMAIN_ATTRIBUTE:
+                    case USER_TASK_INSTANCES_DOMAIN_ATTRIBUTE:
+                        //Skip id, not sortable
+                    case "id":
+                        break;
+                    default:
+                        String typeName;
+                        String name = resolveBaseTypeName(field.getType());
```
Comment (radtriste) [inline data-index/data-index-service/data-index-service-common/src/main/java/org/kie/kogito/index/graphql/query/GraphQLOrderByTypeMapper.java]: why do you need to call `resolveBaseTypeName` whereas `field.getType()` is not of type `GraphQLList` ?

File: data-index/data-index-service/data-index-service-common/src/test/java/org/kie/kogito/index/graphql/JsonPropertyDataFetcherTest.java [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+        //       },
+        //       {"city":"city2",
+        //       "amount":97,
+        //       "approved":false,
+        //       "zipCode":"zipCode2"}
+        //       ],
+        //   "stringArray":["st1","st2"],
+        //   "numberArray":[8,98.6],
+        //   "booleanArray":[true,false]
+        //   }
```
Comment (radtriste) [inline data-index/data-index-service/data-index-service-common/src/test/java/org/kie/kogito/index/graphql/JsonPropertyDataFetcherTest.java]: I would put that as a comment of `getArraysNodeJson` method. wdyt ?

File: data-index/data-index-service/data-index-service-common/src/main/java/org/kie/kogito/index/graphql/JsonPropertyDataFetcher.java (hunk 5 above)
Comment (nmirasch) [reply data-index/data-index-service/data-index-service-common/src/main/java/org/kie/kogito/index/graphql/JsonPropertyDataFetcher.java]: done!

File: data-index/data-index-service/data-index-service-common/src/main/java/org/kie/kogito/index/graphql/query/AbstractInputObjectTypeMapper.java (hunk 6 above)
Comment (nmirasch) [reply data-index/data-index-service/data-index-service-common/src/main/java/org/kie/kogito/index/graphql/query/AbstractInputObjectTypeMapper.java]: Ups, my fault. That was the reason for the baseType variable. Thanks!

File: data-index/data-index-service/data-index-service-common/src/main/java/org/kie/kogito/index/graphql/query/GraphQLOrderByTypeMapper.java (hunk 7 above)
Comment (nmirasch) [reply data-index/data-index-service/data-index-service-common/src/main/java/org/kie/kogito/index/graphql/query/GraphQLOrderByTypeMapper.java]: I agree, I restored old code here to access the type name.

File: data-index/data-index-service/data-index-service-common/src/test/java/org/kie/kogito/index/graphql/JsonPropertyDataFetcherTest.java (hunk 8 above)
Comment (nmirasch) [reply data-index/data-index-service/data-index-service-common/src/test/java/org/kie/kogito/index/graphql/JsonPropertyDataFetcherTest.java]: done!

**Review by radtriste — APPROVED** [review]

**Comment by kie-ci1** [comment]
**(tests) - kogito-apps job** `#16` was: **UNSTABLE**
Possible explanation: This should be test failures


**Test results:**
- PASSED: 2725
- FAILED: 1

Those are the test failures: 
<details>
<summary><a href="https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/kogito/job/main/job/pullrequest/job/kogito-apps.tests.kogito-apps/16/testReport/org.kie.kogito.index.infinispan/ProcessDataIndexInfinispanIT/testProcessInstanceEvents/">org.kie.kogito.index.infinispan.ProcessDataIndexInfinispanIT.testProcessInstanceEvents</a></summary>
java.lang.AssertionError: <br/>1 expectation failed.<br/>JSON path data.UserTaskInstances[0].comments.size() doesn't match.<br/>Expected: is <1><br/>  Actual: <0><br/><br/>
</details>

Please look here: https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/kogito/job/main/job/pullrequest/job/kogito-apps.tests.kogito-apps/16/display/redirect

**Comment by nmirasch** [comment]
jenkins rerun kogito-apps tests

**Comment by cristianonicolai** [comment]
@nmirasch please add to release notes :)

**Pull request merged**